Source/Csla/Core/GraphMerger.cs

```diff
@@ -202,5 +202,173 @@ private void MergeGraph(IEditableCollection target, IEditableCollection source)
+            {
+              await MergeGraphAsync(targetChild, sourceChild);
+            }
+            else
+            {
+              if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceChild);
+              else
+                await Task.Run(() => LoadProperty(target, item, sourceChild));
+            }
+          }
+          else
+          {
+            if (sourceValue is IEditableCollection sourceList)
+            {
+              var targetFieldExists = true;
+              if (targetFieldManager != null && (item.RelationshipType & RelationshipTypes.LazyLoad) == RelationshipTypes.LazyLoad)
+                targetFieldExists = targetFieldManager.FieldExists(item);
+              if (targetFieldExists && ReadProperty(target, item) is IEditableCollection targetList)
+              {
+                await MergeGraphAsync(targetList, sourceList);
+              }
+              else
+              {
+                if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                  Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceList);
+                else
+                  await Task.Run(() => LoadProperty(target, item, sourceList));
+              }
+            }
+            else
+            {
+              if ((item.RelationshipType & RelationshipTypes.PrivateField) == RelationshipTypes.PrivateField)
+                Reflection.MethodCaller.CallPropertySetter(target, item.Name, sourceValue);
+              else if (sourceValue != null || (item.RelationshipType & RelationshipTypes.LazyLoad) != RelationshipTypes.LazyLoad)
+                await Task.Run(() => LoadProperty(target, item, sourceValue));
+            }
+          }
+        }
+        if (source.IsNew)
+        {
+          await Task.Run(() => MarkNew(target));
+        }
+        else if (!source.IsDirty)
+        {
+          await Task.Run(() => MarkOld(target));
+        }
+        else
+        {
+          CopyField(source, target, "_isDirty");
+          CopyField(source, target, "_isNew");
+          CopyField(source, target, "_isDeleted");
+        }
+        await CheckRulesAsync(target);
+      }
+    }
+    /// <summary>
+    /// Merges state from source graph into target graph.
+    /// </summary>
+    /// <param name="target">Target of merge.</param>
+    /// <param name="source">Source for merge.</param>
+    private async Task MergeGraphAsync(IEditableCollection target, IEditableCollection source)
+    {
+      var listType = await Task.Run(() => target.GetType());
+      var childType = await Task.Run(() => Utilities.GetChildItemType(listType));
+      var genericTypeParams = new Type[] { listType, childType };
+      System.Reflection.MethodInfo methodReference;
+      if (typeof(IExtendedBindingList).IsAssignableFrom(listType))
```

Comment:
Why is here reflection necessary to call the methods defined by this type with `target` and `source`?
When reflection is necessary the method infos should be cached within the instance to improve efficency during recursive calls.

Reply:
Reflection is being used in MergeGraph/MergeGraphAsync method to dynamically invoke either the MergeBusinessListGraph/MergeBusinessListGraphAsync or MergeBusinessBindingListGraph/MergeBusinessBindingListGraphAsync method based on the runtime type of the target collection. 

"dynamic" keyword is not supported below .Net 4.x Compatibility projects. Same with delegate.

This will be possible implementation using cache. can you please review so I can commit.

```
    private async Task MergeGraphAsync(IEditableCollection target, IEditableCollection source)
    {
      var listType = target.GetType();
      var childType = Utilities.GetChildItemType(listType);
      var isExtendedBindingList = typeof(IExtendedBindingList).IsAssignableFrom(listType);

      var cacheKey = (listType, childType, isExtendedBindingList);

      if (!_methodCache.TryGetValue(cacheKey, out System.Reflection.MethodInfo methodReference))
      {
        var methodName = isExtendedBindingList ? "MergeBusinessBindingListGraphAsync" : "MergeBusinessListGraphAsync";
        methodReference = GetType().GetMethod(methodName).MakeGenericMethod(listType, childType);
        _methodCache.TryAdd(cacheKey, methodReference);
      }

      var task = (Task)methodReference.Invoke(this, new object[] { target, source });
      await task;
    }
```

Reply:
Fixed. Please check.